File: Minimum_Depth_of_Binary_Tree_bfs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        
        std::queue<std::pair<TreeNode*, int>> q;
        q.push({root, 1});

        while (!q.empty()) {
            TreeNode* node = q.front().first;
            int depth = q.front().second;
            q.pop();

            if (node->left == nullptr && node->right == nullptr) {
                return depth;
            }

            if (node->left != nullptr) {
                q.push({node->left, depth + 1});
            }

            if (node->right != nullptr) {
                q.push({node->right, depth + 1});
            }
        }
        
        return 0;        
    }
};

File: Minimum_Depth_of_Binary_Tree_bfs.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
    if (!root) {
        return 0;
    }
    
    const queue = [];
    queue.push([root, 1]);

    while (queue.length > 0) {
        const [node, depth] = queue.shift();

        if (!node.left && !node.right) {
            return depth;
        }

        if (node.left) {
            queue.push([node.left, depth + 1]);
        }

        if (node.right) {
            queue.push([node.right, depth + 1]);
        }
    }
    
    return 0;    
};
